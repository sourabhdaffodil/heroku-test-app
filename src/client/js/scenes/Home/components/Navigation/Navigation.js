import React, { Component } from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import Icon from '../../../../components/Icon/Icon';
import { media } from '../../../../theme/style-utils';

const RoundedContainer = styled.div`
  border-radius: 3px;
  border: 1px solid ${props => props.theme.color('background')};
  margin: 20px 0;
  border-bottom: 3px solid ${props => props.theme.color('accent')};
  
  ${media.phone`
    margin: 0;  
  `}
`;

const List = styled.ul`
  margin: 0;
  padding: 0;
  display: flex;
`;

const Item = styled.li`
  display: inline-block;
  margin: 0;
  padding: 10px;

  a {
    text-decoration: none;
    color: ${props => props.theme.contrast('light')}
  }

  ${media.phone`
    a {
      display: flex;
      align-items: center;

      span {
        order: 1;
        padding: 5px;
        border-radius: 50%;
        background-color: ${props => props.theme.color('foreground')};
        color: ${props => props.theme.contrast('dark')};
        ${props => props.theme.textSetting('xs')};
        width: 15px;
        height: 15px;
        line-height: 15px;
        text-align: center;
      }
      
      i {
        display: none;
      }
    }
  `}

  .active {
    color: ${props => props.theme.color()};
  }
`;

const Stat = styled.span` 
  font-weight: ${props => props.theme.weight('semiBold')};
  margin: 0 5px;
`;

class Navigation extends Component {
  constructor(props) {
    super(props);
    this.state = {
      activeLink: 0,
      links: [
        {
          url: '/commits',
          icon: 'clock-o',
          stat: 0,
          desc: 'commits',
          active: true,
        },
        {
          url: '/orgs',
          icon: 'users',
          stat: 0,
          desc: 'organizations',
        },
      ],
    };
  }

  componentWillMount() {
    const currentRoute = window.location.href.replace(/.*((:[0-9])*|(.*.com))\//, '/');
    const currentItemIndex = this.state.links.findIndex(item => item.url === currentRoute);

    this.setState({
      activeLink: currentItemIndex >= 0 ? currentItemIndex : 0,
    });
  }

  componentWillReceiveProps(nextProps) {
    this.setState({
      links: [
        {
          url: '/commits',
          icon: 'clock-o',
          stat: nextProps.totalCommits,
          desc: 'commits',
          active: true,
        },
        {
          url: '/orgs',
          icon: 'users',
          stat: nextProps.totalOrganizations,
          desc: 'organizations',
        },
      ],
    });
  }

  setActiveLink(id) {
    this.setState({
      activeLink: id,
    });
  }

  renderLinks() {
    if (this.state.links) {
      return this.state.links.map((item, index) => (
        <Item key={item.url}>
          <Link
            to={item.url}
            className={index === this.state.activeLink ? 'active' : ''}
            onClick={() => this.setActiveLink(index)}
          >
            <Icon icon={item.icon} />
            <Stat>{item.stat}</Stat>
            {item.desc}
          </Link>
        </Item>
      ));
    }

    return null;
  }

  render() {
    return (
      <RoundedContainer>
        <List>
          {this.renderLinks()}
        </List>
      </RoundedContainer>
    );
  }
}

Navigation.propTypes = {
  totalCommits: PropTypes.number.isRequired,
  totalOrganizations: PropTypes.number.isRequired,
};

function mapStateToProps(state) {
  const { commits } = state.history;
  let commitsCounter = 0;
  Object.keys(commits).forEach((key) => {
    commitsCounter += commits[key].length || 0;
  });

  return {
    totalCommits: commitsCounter,
    totalOrganizations: state.organizations.length,
  };
}

export default connect(mapStateToProps)(Navigation);

export {
  Navigation,
  mapStateToProps,
};
