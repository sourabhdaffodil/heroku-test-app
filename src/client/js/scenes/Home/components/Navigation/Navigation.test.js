import React from 'react';
import { shallow } from 'enzyme';
import { Navigation, mapStateToProps } from './Navigation';

describe('Navigation component', () => {
  it('Must render a navigation with the totals', () => {
    const container = shallow(
      <Navigation totalCommits={31} totalOrganizations={10} />,
    );

    expect(container).toMatchSnapshot();
  });

  describe('mapStateToProps function', () => {
    it('Must return an object with no totals', () => {
      const defaultState = {
        history: {
          commits: {},
        },
        organizations: [],
      };
      const result = mapStateToProps(defaultState);
      const expected = {
        totalCommits: 0,
        totalOrganizations: 0,
      };
      expect(result).toEqual(expected);
    });
    it('Must return an object with 6 commits as total', () => {
      const state = {
        history: {
          commits: {
            2015: [
              {},
              {},
              {},
            ],
            2016: [
              {},
            ],
            2017: [
              {},
              {},
            ],
          },
        },
        organizations: [
          {},
          {},
          {},
        ],
      };
      const result = mapStateToProps(state);
      const expected = {
        totalCommits: 6,
        totalOrganizations: 3,
      };
      expect(result).toEqual(expected);
    });
  });
});
