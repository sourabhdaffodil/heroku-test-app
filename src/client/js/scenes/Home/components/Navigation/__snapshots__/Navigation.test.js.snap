// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation component Must render a navigation with the totals 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Navigation
    totalCommits={31}
    totalOrganizations={10}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "class",
    "props": Object {
      "children": <styled.ul>
        <styled.li>
          <Link
            className="active"
            onClick={[Function]}
            replace={false}
            to="/commits"
          >
            <Icon
              icon="clock-o"
            />
            <styled.span>
              0
            </styled.span>
            commits
          </Link>
        </styled.li>
        <styled.li>
          <Link
            className=""
            onClick={[Function]}
            replace={false}
            to="/orgs"
          >
            <Icon
              icon="users"
            />
            <styled.span>
              0
            </styled.span>
            organizations
          </Link>
        </styled.li>
      </styled.ul>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": null,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <styled.li>
            <Link
              className="active"
              onClick={[Function]}
              replace={false}
              to="/commits"
            >
              <Icon
                icon="clock-o"
              />
              <styled.span>
                0
              </styled.span>
              commits
            </Link>
          </styled.li>,
          <styled.li>
            <Link
              className=""
              onClick={[Function]}
              replace={false}
              to="/orgs"
            >
              <Icon
                icon="users"
              />
              <styled.span>
                0
              </styled.span>
              organizations
            </Link>
          </styled.li>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "/commits",
          "nodeType": "class",
          "props": Object {
            "children": <Link
              className="active"
              onClick={[Function]}
              replace={false}
              to="/commits"
            >
              <Icon
                icon="clock-o"
              />
              <styled.span>
                0
              </styled.span>
              commits
            </Link>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": null,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <Icon
                  icon="clock-o"
                />,
                <styled.span>
                  0
                </styled.span>,
                "commits",
              ],
              "className": "active",
              "onClick": [Function],
              "replace": false,
              "to": "/commits",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": null,
                "nodeType": "function",
                "props": Object {
                  "icon": "clock-o",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": null,
                "nodeType": "class",
                "props": Object {
                  "children": 0,
                },
                "ref": null,
                "rendered": 0,
                "type": [Function],
              },
              "commits",
            ],
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "/orgs",
          "nodeType": "class",
          "props": Object {
            "children": <Link
              className=""
              onClick={[Function]}
              replace={false}
              to="/orgs"
            >
              <Icon
                icon="users"
              />
              <styled.span>
                0
              </styled.span>
              organizations
            </Link>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": null,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <Icon
                  icon="users"
                />,
                <styled.span>
                  0
                </styled.span>,
                "organizations",
              ],
              "className": "",
              "onClick": [Function],
              "replace": false,
              "to": "/orgs",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": null,
                "nodeType": "function",
                "props": Object {
                  "icon": "users",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": null,
                "nodeType": "class",
                "props": Object {
                  "children": 0,
                },
                "ref": null,
                "rendered": 0,
                "type": [Function],
              },
              "organizations",
            ],
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "class",
      "props": Object {
        "children": <styled.ul>
          <styled.li>
            <Link
              className="active"
              onClick={[Function]}
              replace={false}
              to="/commits"
            >
              <Icon
                icon="clock-o"
              />
              <styled.span>
                0
              </styled.span>
              commits
            </Link>
          </styled.li>
          <styled.li>
            <Link
              className=""
              onClick={[Function]}
              replace={false}
              to="/orgs"
            >
              <Icon
                icon="users"
              />
              <styled.span>
                0
              </styled.span>
              organizations
            </Link>
          </styled.li>
        </styled.ul>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": null,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <styled.li>
              <Link
                className="active"
                onClick={[Function]}
                replace={false}
                to="/commits"
              >
                <Icon
                  icon="clock-o"
                />
                <styled.span>
                  0
                </styled.span>
                commits
              </Link>
            </styled.li>,
            <styled.li>
              <Link
                className=""
                onClick={[Function]}
                replace={false}
                to="/orgs"
              >
                <Icon
                  icon="users"
                />
                <styled.span>
                  0
                </styled.span>
                organizations
              </Link>
            </styled.li>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "/commits",
            "nodeType": "class",
            "props": Object {
              "children": <Link
                className="active"
                onClick={[Function]}
                replace={false}
                to="/commits"
              >
                <Icon
                  icon="clock-o"
                />
                <styled.span>
                  0
                </styled.span>
                commits
              </Link>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": null,
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  <Icon
                    icon="clock-o"
                  />,
                  <styled.span>
                    0
                  </styled.span>,
                  "commits",
                ],
                "className": "active",
                "onClick": [Function],
                "replace": false,
                "to": "/commits",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": null,
                  "nodeType": "function",
                  "props": Object {
                    "icon": "clock-o",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": null,
                  "nodeType": "class",
                  "props": Object {
                    "children": 0,
                  },
                  "ref": null,
                  "rendered": 0,
                  "type": [Function],
                },
                "commits",
              ],
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "/orgs",
            "nodeType": "class",
            "props": Object {
              "children": <Link
                className=""
                onClick={[Function]}
                replace={false}
                to="/orgs"
              >
                <Icon
                  icon="users"
                />
                <styled.span>
                  0
                </styled.span>
                organizations
              </Link>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": null,
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  <Icon
                    icon="users"
                  />,
                  <styled.span>
                    0
                  </styled.span>,
                  "organizations",
                ],
                "className": "",
                "onClick": [Function],
                "replace": false,
                "to": "/orgs",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": null,
                  "nodeType": "function",
                  "props": Object {
                    "icon": "users",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": null,
                  "nodeType": "class",
                  "props": Object {
                    "children": 0,
                  },
                  "ref": null,
                  "rendered": 0,
                  "type": [Function],
                },
                "organizations",
              ],
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
